### PR TITLE
feat(security): add SafeDir primitive for POSIX-safe filesystem ops

### LINKED ISSUE #266: [security PR2] SafeDir primitive: src/utils/safedir.py
Parent: #264
Branch: `claude/security-symlink-hardening-Nglc6`
Depends on: #265

## Scope

New file: `src/utils/safedir.py`.

```python
class SafeDir:
    """Holds an open dir fd. Every operation uses dir_fd= + O_NOFOLLOW."""

    @classmethod
    def open_root(cls, path: Path) -> "SafeDir": ...

    def open_child(self, name: str, *, flags: int = os.O_RDONLY) -> int: ...
        # O_NOFOLLOW always set; raises if symlink.

    def open_subdir(self, name: str) -> "SafeDir": ...
        # opens with O_DIRECTORY | O_NOFOLLOW

    def scandir(self) -> Iterator[os.DirEntry]: ...
        # via os.scandir(self.fd)

    def lstat(self, name: str) -> os.stat_result: ...
    def unlink(self, name: str) -> None: ...
    def mkdir(self, name: str, mode: int = 0o755) -> None: ...
    def rename_into(self, name: str, other: "SafeDir", other_name: str) -> None: ...

    def open_for_reader(self, name: str) -> int:
        """Open with O_RDONLY|O_NOFOLLOW. Caller wraps via os.fdopen / passes
        as fileobj= to a content reader."""

    def __enter__(self): ...
    def __exit__(self, *a): os.close(self.fd)
```

## Invariants

- Every method takes a single path **component** (not a path). Reject anything containing `/`, `\`, `..`, or NUL — raise `ValueError`.
- Every syscall uses `dir_fd=self.fd`; never reconstructs a string path from attacker-controlled segments.
- `open_*` always sets `O_NOFOLLOW`. On `ELOOP`, raise a typed `SymlinkRejected` exception (subclass of `OSError`).
- `rename_into` uses `os.rename(name, other_name, src_dir_fd=self.fd, dst_dir_fd=other.fd)` — atomic on POSIX.
- File-descriptor lifetime is via context manager only; the class doesn't expose `close()` to discourage leaks.

## Platform

POSIX only. Windows: ship a `_WindowsSafeDir` stub that raises `NotImplementedError` on construction with a clear "POSIX-only hardening, see #264" message. Don't paper over it with a best-effort path-string fallback — that defeats the point.

## Acceptance

- [ ] `src/utils/safedir.py` exists with the API above
- [ ] `tests/utils/test_safedir.py` covers:
  - successful open / scandir / read of a regular file
  - `SymlinkRejected` raised on a symlinked child (file and directory)
  - `ValueError` raised on `name` containing `/`, `..`, NUL
  - `rename_into` succeeds same-dir and cross-dir
  - `unlink` via `dir_fd=`
  - context manager closes the fd (verify via `/proc/self/fd` count)
  - POSIX-skip on Windows
- [ ] No call-site changes in `src/` (just the new module + tests)
- [ ] `tests/ci/test_symlink_safety_lints.py` recognises calls routed through `SafeDir` as compliant

### PR DESCRIPTION
**Pull request by curdriceaurora**
Closes #266. Second PR in the security hardening series tracked under #264.

## Summary

`src/utils/safedir.py` introduces the `SafeDir` primitive that holds an open directory file descriptor and routes every filesystem operation through `dir_fd=` + `O_NOFOLLOW`. PRs 3–6 will thread it through the read-side ingestion, dedupe, undo, and watcher paths.

### Design invariants enforced by the implementation

- **Every method takes a single path *component*, never a path.** Names containing `/`, `\`, `..`, `.`, NUL, or empty strings raise `ValueError` before any syscall. Attacker-controlled segments (file content, AI output, watcher event payloads) cannot escape the held directory.
- **`O_NOFOLLOW` is always set on opens.** Linux quirk: with `O_DIRECTORY | O_NOFOLLOW` a symlink-to-dir returns `ENOTDIR` rather than `ELOOP`, so the open helpers do a post-error `lstat` to disambiguate and raise `SymlinkRejected` (subclass of `OSError`) in either case.
- **No path-string reconstruction.** Every syscall uses `dir_fd=self.fd` with the bare component.
- **fd lifetime is via context manager only**; double-exit is harmless. The class deliberately does not expose a public `close()` to discourage leaks.

### API surface

- `SafeDir.open_root(path)` — factory
- `open_child(name, *, flags=O_RDONLY)` — generic open (always `O_NOFOLLOW`)
- `open_for_reader(name)` — canonical "pass to content reader" helper; pair with `os.fdopen(fd, "rb")`
- `open_subdir(name)` — returns a new `SafeDir` for a child directory
- `scandir()` — `os.scandir(self.fd)`
- `lstat(name)` — for inode-pinning workflows (PR4 / PR5)
- `mkdir`, `unlink`, `rename_into(name, other, other_name)`
- Context-manager protocol

### Platform

POSIX only. `open_root` raises `NotImplementedError` on Windows; companion tests skip there. Windows port deferred per #264.

## Test plan

- [x] `tests/utils/test_safedir.py` — **48 cases**, all pass
  - Name validation across every method (parametrized over `/`, `..`, `.`, `""`, NUL, `\`, absolute paths)
  - Happy paths for every operation (open, scandir, lstat, unlink, mkdir, `rename_into` same-dir + cross-dir, `open_subdir` chaining)
  - Symlink rejection on file open, dir open, and root open
  - Predicate-T10 negatives — entries named "link" / "documents_link" that are actually regular files/dirs must succeed
  - fd lifetime — 20-iteration loop verifies no leak via `/proc/self/fd`
  - Missing-entry semantics — `FileNotFoundError`, not generic `OSError`
- [x] CI subset (`pytest -m ci`) — 4400 passed, 7 skipped, plus the 2 pre-existing unrelated failures confirmed from PR #271
- [x] `ruff check` + `ruff format --check` clean
- [x] Advisory baseline of `scripts/check_safedir_required.py` unchanged at 48 call sites — `src/utils/safedir.py` is already allowlisted (set up in PR #271)

## What this is NOT

This PR adds the primitive but does not migrate any existing call sites. None of the 48 advisory-mode rail violations are resolved here — those come down in:

- #267 (PR3) — Read-side ingestion migration (closes LLM-exfiltration vector)
- #268 (PR4) — Inode-pin dedupe
- #269 (PR5) — History `(dev, ino)` + undo verification
- #270 (PR6) — Watcher + move-destination hardening

https://claude.ai/code/session_01EnPpeovvetPSR8Vwhxsn9j

---
_Generated by [Claude Code](https://claude.ai/code/session_01EnPpeovvetPSR8Vwhxsn9j)_

## Summary by CodeRabbit

* **New Features**
  * Added secure directory operations with automatic symlink attack prevention for POSIX systems.

* **Improvements**
  * Enhanced error handling for file operations with improved support for permission-sensitive scenarios.
  * Expanded test coverage for edge cases and cross-platform conditions.

[Review Change Stack](https://app.coderabbit.ai/change-stack/curdriceaurora/fo-core/pull/272?utm_source=github_walkthrough&utm_medium=github&utm_campaign=change_stack)

### CODE REVIEW

File: src/utils/safedir.py [hunk 1]
```diff
@@ -0,0 +1,360 @@
...
+        """Release the held directory fd. Idempotent on double-exit.
+
+        Closing an already-closed fd would raise ``EBADF``, which is
+        noise in failure-path teardown; suppress it.
+        """
+        if self._closed:
+            return
+        self._closed = True
+        try:
+            os.close(self._fd)
```
[the remark this answers is not in the document]
Comment (curdriceaurora) [reply src/utils/safedir.py]: Already addressed in commit 06485f1 (pushed just now — the review was on the prior commit 5c40ae4 before this landed).

Implementation choices vs your suggested patch:

- Used `_check_open()` (returns `None`, raises `ValueError`) rather than `_require_open()` returning the fd. Functionally equivalent; I prefer a guard that doesn't entangle return-flow with validity. Every method calls `self._check_open()` at the top and then uses `self._fd` as before.
- Used `ValueError("SafeDir has been closed; cannot use after context-manager exit")` rather than `OSError(EBADF, ...)`. Two reasons: (1) the failure isn't really an OS error — the OS would happily operate on the recycled fd, which is exactly the bug we're avoiding; (2) `ValueError` aligns with how this class signals other misuse (`_validate_name` raises ValueError for bad name components).
- Added defense-in-depth: `__exit__` sets `self._fd = -1` after `os.close`. Even if `_check_open` is somehow bypassed (e.g. direct attribute access `sd._fd`), any syscall targeting -1 fails immediately with EBADF rather than silently operating on a recycled fd.
- `rename_into` checks both ends (`self._check_open()` and `other._check_open()`).
- `__enter__` deliberately doesn't guard — re-entering a closed `SafeDir` would already fail on the first method call inside the `with`, which is the actionable failure point. Guarding `__enter__` only would defer the error to the first method anyway.

Tests: added `TestUseAfterClose` (11 cases) exercising fd-property + every public method + cross-dir rename_into + the `_fd == -1` invariant.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: src/utils/safedir.py [hunk 2]
```diff
@@ -0,0 +1,360 @@
...
+        except OSError:  # pragma: no cover - defensive
+            pass
```
[the remark this answers is not in the document]
Comment (curdriceaurora) [reply src/utils/safedir.py:281]: Addressed in commit 06485f1 — this is a duplicate of the Codex P2 finding on r3254858073 (same root cause, same scope). The PR review here was on `5c40ae4` before my fix landed; CodeRabbit's other thread (`test_safedir.py:312`) already auto-resolved after seeing 06485f1, so I expect this one will too on the next pass.

Implementation differences from your suggested diff:

- `_check_open()` (returns `None`, raises) rather than `_require_open()` (returns the fd). Functionally equivalent; the side-effect-only guard reads cleaner at call sites (`self._check_open()` on its own line, then `self._fd` as before — no unrelated binding to thread through every syscall site).
- `ValueError("...closed...")` rather than `OSError(EBADF, ...)`. The failure isn't an OS error — the OS would happily operate on the recycled fd, which is the bug we're preventing. `ValueError` matches how the class already signals misuse (`_validate_name`).
- `__enter__` deliberately doesn't guard. Re-entering a closed SafeDir would fail on the first method call inside the `with` anyway, which is the actionable failure point. Adding a guard to `__enter__` would only shift the same error one frame earlier.
- Defense-in-depth: `__exit__` sets `self._fd = -1` after `os.close`. Even if `_check_open` is bypassed (e.g. via direct attribute access), syscalls targeting -1 fail loudly with EBADF rather than silently operating on a recycled fd.
- `rename_into` checks both ends.

Tests in the same commit: `TestUseAfterClose` (11 cases) — fd-property + every public method + cross-dir `rename_into` + the `_fd == -1` invariant.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: src/utils/safedir.py [hunk 3]
```diff
@@ -0,0 +1,396 @@
...
+        operations (the SafeDir invariant still holds: the open is
+        anchored to ``self.fd`` and never follows a symlink).
+
+        Returns the opened fd. Caller is responsible for closing it
+        (use ``os.fdopen`` / ``os.close`` / ``contextlib.closing``).
+        """
+        self._check_open()
+        _validate_name(name)
+        try:
+            return os.open(name, flags | os.O_NOFOLLOW, dir_fd=self._fd)
```
[the remark this answers is not in the document]
Comment (curdriceaurora) [reply src/utils/safedir.py]: Applied in commit ba08e21.

Approach: reject `O_PATH` at the `open_child` API boundary rather than `fstat`-after-open. Reasoning:

1. `open_child`'s documented contract is "if the entry is a symlink, raises SymlinkRejected". Letting `O_PATH` through and post-checking via `fstat(S_ISLNK)` would silently violate the contract for callers who passed `O_PATH` thinking they had a normal handle — they'd get a `SymlinkRejected` they couldn't have anticipated from the flag they actually used.
2. SafeDir's current API surface doesn't need `O_PATH`. `open_for_reader` is `O_RDONLY`; `open_subdir` is `O_RDONLY | O_DIRECTORY`. Today there's no exploitable path — but the contract violation makes future hardening fragile.
3. If PR4/PR5 genuinely need `O_PATH` semantics for inode-pinning, they can ship a dedicated method (`open_for_stat`?) that handles the post-open `S_ISLNK` check explicitly.

Platform-portability: `_O_PATH = getattr(os, "O_PATH", 0)` — on macOS/BSD where `O_PATH` doesn't exist, the guard is a no-op (the bug doesn't exist there either, so nothing to reject).

Tests added (61 total now):
- `test_open_child_rejects_o_path` — `open_child(name, flags=os.O_PATH)` raises `ValueError`
- `test_open_child_rejects_o_path_even_for_real_files` (T10 predicate-negative) — the rejection fires on the flag, not on whether the entry is a symlink; a regular file with `O_PATH` is refused too, so callers can't "just happen to" use the flag safely and later trip with an attacker-controlled symlink

---
_Generated by [Claude Code](https://claude.ai/code)_

File: src/utils/safedir.py [hunk 4]
```diff
@@ -0,0 +1,419 @@
...
+        _validate_name(name)
+        if _O_PATH and (flags & _O_PATH):
+            raise ValueError(
+                "open_child does not support O_PATH: O_PATH|O_NOFOLLOW "
+                "would return an fd for the symlink instead of refusing it"
+            )
+        try:
+            return os.open(name, flags | os.O_NOFOLLOW, dir_fd=self._fd)
+        except OSError as exc:
+            raise _wrap_open_errno(exc, name) from exc
```
[the remark this answers is not in the document]
Comment (curdriceaurora) [reply src/utils/safedir.py]: Applied in commit 98d70a2.

Confirmed the issue locally: `os.open("symlink", O_WRONLY | O_CREAT | O_EXCL | O_NOFOLLOW)` returns `EEXIST` rather than `ELOOP` on Linux — `O_EXCL` fires before the `O_NOFOLLOW` path. So `open_child` was returning `FileExistsError` for symlinked targets, leaking through the documented "symlink → SymlinkRejected" contract.

Fix follows the same disambiguation pattern as `open_subdir`'s `ENOTDIR` handling: on `EEXIST`, `lstat` the entry; if it's a symlink, raise `SymlinkRejected`. Tolerable one-syscall TOCTOU between the failed open and the lstat — the open already refused to dereference, and a same-name swap-in just confirms the entry IS now a symlink, which is still the security-relevant case worth raising.

Tests added (64 total now):
- `test_open_child_excl_create_rejects_symlinked_target` — real symlink → `SymlinkRejected` (not `FileExistsError`)
- `test_open_child_excl_create_passes_through_real_file_eexist` (T10 predicate-negative) — regular existing file → `FileExistsError` still surfaces; disambiguation fires only on actual symlinks
- `test_open_child_excl_create_succeeds_for_new_name` (positive control) — non-existent name → exclusive-create succeeds, no regression on the happy path

---
_Generated by [Claude Code](https://claude.ai/code)_

File: src/utils/safedir.py [hunk 5]
```diff
@@ -0,0 +1,438 @@
...
+            )
+        flags = os.O_RDONLY | os.O_DIRECTORY | os.O_NOFOLLOW
+        try:
+            fd = os.open(str(path), flags)
+        except OSError as exc:
+            # ``O_DIRECTORY | O_NOFOLLOW`` against a symlink-to-dir returns
+            # ENOTDIR on Linux (the symlink inode isn't a directory).
+            # Disambiguate via lstat — symlink → SymlinkRejected, anything
+            # else → propagate.
+            if exc.errno == errno.ELOOP or (exc.errno == errno.ENOTDIR and path.is_symlink()):
```
[the remark this answers is not in the document]
Comment (curdriceaurora) [reply src/utils/safedir.py:217]: Applied in commit 6fb63eb.

Confirmed: with the previous code, `SafeDir.open_root(str(symlinked_path))` would raise `AttributeError: 'str' object has no attribute 'is_symlink'` instead of the documented `SymlinkRejected`. Type hints don't enforce runtime types in Python, and `str(path)` is idempotent for both `str` and `Path` so the `os.open` call earlier in the function silently accepted the wrong type.

Two changes:

1. Widened the type hint to `str | os.PathLike[str]` — the runtime contract is now honest. `os.PathLike[str]` covers `pathlib.Path` and any other path-like (CLI/config callers, third-party path wrappers).
2. Normalized at the boundary: `path = Path(path)` at function entry. Idempotent for existing `Path` callers; converts str/PathLike once so subsequent uses (`str(path)`, `path.is_symlink()`, etc.) all work uniformly without sprinkling defensive conversions through the body.

Test added: `test_open_root_rejects_symlinked_root_when_passed_as_string` — exercises the previous failure mode (string-path caller on a symlinked root) and asserts `SymlinkRejected`. 65 tests total.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: src/utils/safedir.py [hunk 6]
```diff
@@ -0,0 +1,445 @@
...
+            # ELOOP is the direct symlink-rejection path (O_NOFOLLOW on a
+            # symlink with non-create flags). EEXIST shadows it for the
+            # O_EXCL case — disambiguate via lstat. Tolerating the
+            # one-syscall TOCTOU between failed open and lstat is fine:
+            # the open already didn't follow a symlink, and a same-name
+            # swap-in just means the entry IS now a symlink, which is
+            # still the security-relevant case.
+            if exc.errno == errno.ELOOP or (
+                exc.errno == errno.EEXIST and _is_symlink_at(name, self._fd)
+            ):
```
[the remark this answers is not in the document]
Comment (curdriceaurora) [reply src/utils/safedir.py]: Applied in commit 5338d25.

This completes the errno-shadow disambiguation for `open_child`. Three flag-combinations can shadow the documented "symlink → SymlinkRejected" contract because their respective error paths fire before `O_NOFOLLOW`'s `ELOOP`:

| Flag combo | Errno on symlink | Now handled |
|------------|------------------|-------------|
| `O_RDONLY \| O_NOFOLLOW` | `ELOOP` (direct) | ✓ original |
| `O_CREAT \| O_EXCL` | `EEXIST` | ✓ commit 98d70a2 |
| `O_DIRECTORY` | `ENOTDIR` | ✓ this commit |

Consolidated the `EEXIST` and `ENOTDIR` branches into a single `shadowed_by_symlink = exc.errno in (EEXIST, ENOTDIR) and _is_symlink_at(...)` check — same lstat path, just a wider errno set. Disambiguation fires only on actual symlinks; legitimate non-symlink causes (regular file with `O_DIRECTORY`, existing file with `O_EXCL`) still surface as `NotADirectoryError` / `FileExistsError`.

Also updated the docstring to enumerate all three cases so future readers don't have to reverse-engineer the flag-vs-errno interactions.

Tests (67 total):
- `test_open_child_directory_flag_rejects_symlinked_target` — symlink + `O_DIRECTORY` → `SymlinkRejected`
- `test_open_child_directory_flag_propagates_enotdir_for_regular_file` (T10 predicate-negative) — regular file + `O_DIRECTORY` → still `NotADirectoryError`

---
_Generated by [Claude Code](https://claude.ai/code)_

File: src/utils/safedir.py [hunk 7]
```diff
@@ -0,0 +1,457 @@
...
+    def scandir(self) -> Iterator[os.DirEntry[str]]:
+        """Iterate over entries in this directory, yielding ``os.DirEntry``s.
+
+        The underlying ``os.scandir(self.fd)`` reads the directory by
+        file descriptor — there is no path-string lookup that could be
+        intercepted between this call and a follow-up operation on the
+        same name.
+        """
+        self._check_open()
+        return iter(os.scandir(self._fd))
```
[the remark this answers is not in the document]
Comment (curdriceaurora) [reply src/utils/safedir.py]: Applied in commit f204f53.

Confirmed the footgun: a caller writing `for entry in sd.scandir(): if entry.is_file(): ...` would silently classify a symlink-to-regular-file as a file (because `DirEntry.is_file()` defaults to `follow_symlinks=True`). Same for `.is_dir()` and `.stat()`. Even before reaching `open_for_reader`, the caller could leak metadata about the target via `entry.stat().st_size` etc.

Approach: filter symlinks at the source. SafeDir's whole purpose is "this tree contains only the things we control" — symlinks in there are by definition unsafe (the symlink rejection elsewhere already establishes that). Filtering at scandir matches `safe_walk`'s contract and removes the footgun entirely. Callers who legitimately need to inspect a symlink entry can call `lstat(name)` directly — that's the right primitive for "tell me what's at this name without following".

Two changes:

1. `scandir` is now a generator that skips entries where `DirEntry.is_symlink()` returns True (`is_symlink()` itself never follows, so the check is safe). Per-entry stat failures are skipped too — defensive stance matches `safe_walk`.
2. Wrapped the underlying iterator in `with os.scandir(...) as it:`. The previous `iter(os.scandir(self._fd))` returned the iterator without entering its context manager, leaking the underlying handle until GC. The generator now closes deterministically even if the caller exhausts only part of it.

Test: `test_scandir_filters_symlinks` — directory with one regular file plus a file-symlink and a directory-symlink; only the regular file is yielded. Defence check: every yielded entry passes `not entry.is_symlink()`. 68 tests total.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: src/utils/safedir.py [hunk 8]
```diff
@@ -0,0 +1,478 @@
...
+        Caller is responsible for ensuring ``other`` is on the same
+        filesystem; cross-filesystem renames will raise ``OSError``
+        (``EXDEV``) and the caller must fall back to copy + unlink (see
+        ``undo/durable_move.py`` for the existing pattern).
+        """
+        self._check_open()
+        other._check_open()
+        _validate_name(name)
+        _validate_name(other_name)
+        os.rename(name, other_name, src_dir_fd=self._fd, dst_dir_fd=other._fd)
```
[the remark this answers is not in the document]
Comment (curdriceaurora) [reply src/utils/safedir.py:529]: Applied in commit e068afa.

You're right — `os.rename` has no `O_NOFOLLOW` and Linux offers no atomic rename-if-not-symlink primitive (renameat2 has RENAME_NOREPLACE etc. but nothing symlink-aware). Mitigation is a pre-rename `lstat` check via the existing `_is_symlink_at` helper.

This narrows the attack window from caller-dependent ("scandir-to-rename" — seconds or minutes while AI classification runs between enumeration and move) down to a single syscall ("lstat-to-rename"). Residual race is acknowledged in the docstring as the best we can do without kernel support.

I only checked the source side. Reasoning for skipping the destination:
- `os.rename` with `dst_dir_fd=other._fd` ensures the rename target is resolved through the held destination dir fd — the parent directory itself can't be a symlink at this point (`open_root` / `open_subdir` already refused).
- If the destination *entry* exists as a symlink, `os.rename` overwrites it (Linux default for rename). That's not contamination — the attacker's symlink gets replaced by our trusted source.

Tests (69 total):
- `test_rename_into_rejects_symlinked_source` — pre-existing symlink at source → `SymlinkRejected`; destination remains empty; source symlink stays in place (verifying the rename was refused, not partially executed).

---
_Generated by [Claude Code](https://claude.ai/code)_

File: src/utils/safedir.py [hunk 9]
```diff
@@ -0,0 +1,493 @@
...
+        """
+        self._check_open()
+        _validate_name(name)
+        if _O_PATH and (flags & _O_PATH):
+            raise ValueError(
+                "open_child does not support O_PATH: O_PATH|O_NOFOLLOW "
+                "would return an fd for the symlink instead of refusing it"
+            )
+        try:
+            return os.open(name, flags | os.O_NOFOLLOW, dir_fd=self._fd)
```
[the remark this answers is not in the document]
Comment (curdriceaurora) [reply src/utils/safedir.py]: Applied in commit 18b190c.

Real bug — confirmed locally: `os.open(name, O_WRONLY|O_CREAT|O_EXCL, dir_fd=fd)` with no `mode=` defaults to `0o777`. Under the test container's `0o022` umask the file landed as `0o755`. SafeDir is meant for data files (text, images, output of fo organize), so executable bits make no sense.

Fix: add `mode: int = 0o666` parameter, mirroring the high-level `open()` builtin (which is what most callers expect when they think "default file permissions"). Passes through to `os.open`, kernel still applies umask, so default umask `022` gives `0o644` — non-executable.

Callers that need stricter permissions (e.g. embedding-cache files with secrets, credential stores in PR4/PR5) can pass `mode=0o600` explicitly.

Note `mode` is only consulted when `O_CREAT` is in `flags` (POSIX `open(2)` semantics), so this is a no-op for `open_for_reader` and other read-only callers — no behavioral change to the dominant code path.

Tests (71 total):
- `test_open_child_creates_files_with_non_executable_mode` — default-mode create never produces execute bits, regardless of umask
- `test_open_child_honors_explicit_mode` — `mode=0o600` with umask 0 produces exactly `0o600`

---
_Generated by [Claude Code](https://claude.ai/code)_

File: src/utils/safedir.py [hunk 10]
```diff
@@ -0,0 +1,501 @@
...
+            for entry in it:
+                try:
+                    if entry.is_symlink():
+                        continue
+                except OSError:
+                    # Per-entry stat failure (race with deletion,
+                    # permission flip, etc.) — skip defensively. Matches
+                    # safe_walk's stance.
+                    continue
+                yield entry
```
[the remark this answers is not in the document]
Comment (curdriceaurora) [reply src/utils/safedir.py]: Applied in commit 9ff7eb1.

You're right — the previous filter was a half-measure. Even with symlinks skipped at scan time, exposing `DirEntry` left callers free to reach `.is_file()` / `.is_dir()` / `.stat()`, all of which default to `follow_symlinks=True`. A post-scan TOCTOU swap (regular file → symlink) means a downstream `entry.stat()` follows the now-symlink and returns attacker-target metadata.

Fix: yield plain `str` names instead of `DirEntry`. Callers who need any info beyond the name use SafeDir's own primitives:

- `self.lstat(name)` for metadata (always `follow_symlinks=False`)
- `self.open_for_reader(name)` for reads (`O_NOFOLLOW`)
- `self.open_subdir(name)` for descent (`O_DIRECTORY | O_NOFOLLOW`)

Every one of those does an atomically-correct check against the entry's *current* state, not whatever was cached at scan time. The API surface no longer exposes any symlink-following method to callers.

The cached `is_symlink()` filter stays — it's a useful hint to skip entries that were symlinks at scan time, even if the cached `d_type` isn't atomically correct against post-scan swaps. Atomic safety comes from the downstream checks; the filter just prunes the obvious cases so the caller doesn't have to.

Breaking change to the (yet-unshipped) SafeDir API: `scandir()` now returns `Iterator[str]` instead of `Iterator[os.DirEntry[str]]`. Acceptable because the module is new in this PR and not yet threaded through any callers (PR3+ will be the first consumers; they'll see only the new signature).

Tests (71 total) updated to match the new signature.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: src/utils/safedir.py [hunk 11]
```diff
@@ -0,0 +1,516 @@
...
+            for entry in it:
+                try:
+                    if entry.is_symlink():
+                        continue
+                except OSError:
+                    # Per-entry stat failure (race with deletion,
+                    # permission flip, etc.) — skip defensively. Matches
+                    # safe_walk's stance.
+                    continue
+                yield entry.name
```
[the remark this answers is not in the document]
Comment (curdriceaurora) [reply src/utils/safedir.py]: Applied in commit 8559299.

You're right — confirmed via CPython source: `os.scandir(fd)` calls `fdopendir(dup(fd))`, so the `ScandirIterator` owns a duplicated handle. After `SafeDir.__exit__` closes the SafeDir's own fd, the iterator's dup'd fd is still live and the generator can keep yielding.

Fix: materialize names eagerly into a list, then return `iter(names)`. The `with os.scandir(...) as it:` block closes the dup'd fd before scandir() returns. Caller receives a plain in-memory iterator that holds no OS handles — safe to consume after `SafeDir.__exit__` (the data is already materialized), but no syscalls happen.

Considered the lazy-with-_closed-check alternative but eager materialization is cleaner:
- The `with` block closes the dup'd fd deterministically.
- No race between `_closed` flag write and `_check_open()` read inside the generator.
- The caller's behavior is consistent: scandir() either succeeds fully or raises at startup; no partial yields followed by a surprise stop.

Tradeoff: O(N) memory in directory size. For typical SafeDir trees (user organize roots, capped by `safe_walk`-style limits in the caller code) the cost is negligible (~50 bytes × N names = ~5MB for 100K entries).

Tests (72 total):
- `test_scandir_materializes_eagerly_so_iteration_cannot_outlive_safedir` — consume the iterator after `SafeDir.__exit__` and confirm the pre-exit data still yields correctly with no leaked fds and no syscalls.

---
_Generated by [Claude Code](https://claude.ai/code)_
